File: frontend/lib/schemas/dpa.ts
import { DocSchema } from "../doc-schema";

export const dpaSchema: DocSchema = {
  id: "dpa",
  title: "Data Processing Agreement",
  subtitle: "Cover Page",
  infoBox:
    'This Data Processing Agreement (the "DPA") consists of: (1) this Cover Page and (2) the Common Paper DPA Standard Terms Version 1.0 identical to those posted at commonpaper.com/standards/data-processing-agreement/1.0. The DPA governs how personal data is processed in compliance with data protection regulations such as GDPR.',
  coverFields: [
    {
      key: "effectiveDate",
      label: "Effective Date",
      type: "date",
      placeholder: "Select a date",
    },
    {
      key: "processingPurpose",
      label: "Nature and Purpose of Processing",
      hint: "Why personal data is being processed",
      type: "textarea",
      placeholder: "Describe the processing purpose...",
    },
    {
      key: "personalDataTypes",
      label: "Categories of Personal Data",
      hint: "Types of personal data being processed",
      type: "textarea",
      placeholder: "e.g. Name, email, IP address, usage data",
    },
    {
      key: "dataSubjectCategories",
      label: "Categories of Data Subjects",
      hint: "Who the personal data relates to",
      type: "text",
      placeholder: "e.g. End users, employees, customers",
    },
    {
      key: "governingLaw",
      label: "Governing Law",
      type: "text",
      placeholder: "State or EU Member State",
    },
    {
      key: "chosenCourts",
      label: "Chosen Courts",
      type: "text",
      placeholder: "City/county and state",
    },
  ],
  parties: [
    { label: "Controller", fields: ["name", "title", "company", "noticeAddress", "date"] },
    { label: "Processor", fields: ["name", "title", "company", "noticeAddress", "date"] },
  ],
  signingNote: "By signing this Cover Page, each party agrees to enter into this Data Processing Agreement as of the Effective Date.",
  templateFile: "templates/DPA.md",
  footerNote: "Common Paper Data Processing Agreement (Version 1.0) free to use under CC BY 4.0.",
  greeting:
    "Hi! I'm your Data Processing Agreement assistant. I'll help you fill in the DPA Cover Page through conversation. This governs how personal data is processed in compliance with data protection regulations like GDPR. Let's start -- what is the purpose of processing personal data under this agreement? (e.g. 'Providing cloud-based CRM services', 'Processing payroll for employees', 'Delivering email marketing campaigns')",
  pdfFilename: "data-processing-agreement.pdf",
};

File: backend/doc_configs/dpa.py
from typing import Optional
from pydantic import BaseModel
from doc_configs.shared import PartialPartyInfo


SYSTEM_PROMPT = """You are a legal document assistant helping a user fill in a Data Processing Agreement (DPA) Cover Page.

## Fields you must collect

1. **Effective Date** -- when the agreement takes effect
   Ask: "When should this DPA take effect? (e.g. 'today', 'April 1 2026', 'next Monday')"
   For shortcuts like "today", "tomorrow", "Mon", "next Mon", use this lookup table:
{date_lookup}
   For full dates (e.g. "March 15 2026", "2026-03-15"), convert to YYYY-MM-DD format directly.

2. **Nature and Purpose of Processing** -- why personal data is being processed
   Ask: "What is the purpose of processing personal data? (e.g. 'Providing cloud-based CRM services', 'Processing payroll for employees', 'Delivering email marketing campaigns')"

3. **Categories of Personal Data** -- types of personal data being processed
   Ask: "What types of personal data will be processed? (e.g. 'Name, email, IP address, usage data', 'Employee records, payroll information', 'Contact details, purchase history')"

4. **Categories of Data Subjects** -- who the personal data relates to
   Ask: "Whose personal data will be processed? (e.g. 'End users', 'Employees', 'Customers and their contacts')"

5. **Governing Law** -- jurisdiction whose laws govern the agreement
   Ask: "Which jurisdiction's law should govern this DPA? (e.g. 'California', 'Germany', 'Ireland')"

6. **Chosen Courts** -- which courts handle disputes
   Ask: "Which courts should have jurisdiction over disputes? (e.g. 'San Francisco, California', 'Berlin, Germany', 'Dublin, Ireland')"

7. **Party 1 (Controller)** -- name, title, company, notice address, signing date
   Ask for all sub-fields. Example: "Jane Smith, DPO of DataCo, jane@dataco.com, signing today"

8. **Party 2 (Processor)** -- same sub-fields as Party 1

## Current document state
{current_fields}

## Instructions
- Work through the fields in order, but adapt naturally to what the user already told you
- Ask 1-2 questions per turn maximum
- Always include a concrete example in parentheses when asking a question
- When the user answers, confirm what you captured then always end with a follow-up question
- Once all fields are collected, tell the user the document is complete and they can download the PDF

## Clearing fields
If the user asks to clear fields (e.g. "clear all", "reset", "clear the date"):
- To clear a specific field, return that field set to its empty value and confirm what was cleared
- To clear all fields, return every field reset to empty values, confirm, then ask the first question again
- Never clear silently

Return JSON with:
- "message": your conversational reply (include examples when asking questions)
- "fields": only the fields you are confident about from this turn; omit everything else
"""


class PartialFields(BaseModel):
    effectiveDate: Optional[str] = None
    processingPurpose: Optional[str] = None
    personalDataTypes: Optional[str] = None
    dataSubjectCategories: Optional[str] = None
    governingLaw: Optional[str] = None
    chosenCourts: Optional[str] = None
    party1: Optional[PartialPartyInfo] = None
    party2: Optional[PartialPartyInfo] = None


class AiResponse(BaseModel):
    message: str
    fields: PartialFields
